File: 125-valid-palindrome/valid-palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        return iterativeSolution(s);
    }

    boolean iterativeSolution(String s) {
        if (s.isEmpty()) {
            return true;
        }

        int l = 0;
        int r = s.length() - 1;

        while (l < r) {
            while(l < r && ! Character.isLetterOrDigit(s.charAt(l))) {
                l++;
            }

            while (l < r && ! Character.isLetterOrDigit(s.charAt(r))) {
                r--;
            }

            if (Character.toLowerCase(s.charAt(l)) != Character.toLowerCase(s.charAt(r))) {
                return false;
            }
            l++;
            r--;
        }

        return true;

    }
}

File: 125-valid-palindrome/valid-palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        l, r = 0, len(s) - 1

        def isAlphaNumeric(c):
            return (
                ord('0') <= ord(c) <= ord('9') or
                ord('a') <= ord(c) <= ord('z') or
                ord('A') <= ord(c) <= ord('Z')
            )
        
        while l < r:
            while not isAlphaNumeric(s[l]) and l < r:
                l += 1
            
            while not isAlphaNumeric(s[r]) and l < r:
                r -= 1
            
            if s[l].lower() != s[r].lower():
                return False
            
            l += 1
            r -= 1
        
        return True

                